google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb

```diff
@@ -123,6 +123,86 @@ def empty?
           fields.empty?
         end
 
+        ##
+        # Load the schema from a JSON file.
+        #
+        # The JSON schema file is the same as for the [`bq`
+        # CLI](https://cloud.google.com/bigquery/docs/schemas#specifying_a_json_schema_file)
+        # consisting of an array of JSON objects containing the following:
+        # - `name`: The column's [data
+        #   type](https://cloud.google.com/bigquery/docs/schemas#standard_sql_data_types)
+        # - `type`: The column [name](https://cloud.google.com/bigquery/docs/schemas#column_names)
+        # - `description`: (Optional) The column's [description](https://cloud.google.com/bigquery/docs/schemas#column_descriptions)
+        # - `mode`: (Optional) The column's [mode](https://cloud.google.com/bigquery/docs/schemas#modes)
+        #   (if unspecified, mode defaults to `NULLABLE`)
+        # - `fields`: If `type` is `RECORD`, an array of objects defining child
+        #   fields with these properties
+        #
+        # @param [IO, String, Array<Hash>] source An `IO` containing the JSON
+        #   schema, a `String` containing the JSON schema, or an `Array` of
+        #   `Hash`es containing the schema details.
+        #
+        # @return [Schema] The schema so that commands are chainable.
+        #
+        # @example
+        #   require "google/cloud/bigquery"
+        #
+        #   bigquery = Google::Cloud::Bigquery.new
+        #   dataset = bigquery.dataset "my_dataset"
+        #   table = dataset.table "my_table" do |table|
+        #     table.schema.load "path/to/schema.json"
+        #   end
+        #
+        def load source
+          if source.respond_to?(:rewind) && source.respond_to?(:read)
+            source.rewind
+            schema_json = String source.read
+          elsif source.is_a? Array
+            schema_json = source.to_json
+          else
+            schema_json = String source
+          end
+
+          schema_json = migrate_json schema_json
+
+          raise ArgumentError unless schema_json
+
+          @original_json = schema_json
```

Comment:
It's set in [`::from_gapi`](https://github.com/keylimetoolbox/google-cloud-ruby/blob/a07c15fb43160b4ddae2617ddacd94c9ed116ba8/google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb#L407). I couldn't use that function directly, because I already had an instance I needed to change — I didn't want to create a new instance.

It looks like this is used by the [`#changed?` method](https://github.com/keylimetoolbox/google-cloud-ruby/blob/a07c15fb43160b4ddae2617ddacd94c9ed116ba8/google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb#L395).

Should I not be setting it here? Is this considered a change?

Reply:
Ah, that makes sense now. If you call load on a schema should `changed?` be `true` until it is saved? @quartzmo, what do you think about this? What state should the schema object be in after calling `load`?

Reply:
It looks to me like calling `@original_json = schema_json` will result in `changed?` returning `false`, because `@original_json` will be the same as `@gapi.to_json`. If `changed?` is `false`, won't `Table#check_for_mutated_schema!` fail to persist the change? I might be wrong (still need to look for tests), but it seems to me this statement should be removed.

Reply:
That’s my thinking. I don’t think `load` should be setting `@original_json`.

Reply:
Remove in fa42191

Reply:
The PR needs both acceptance and unit tests for using `load` within a block passed to `Table#schema` for update.